### PR TITLE
Add kind to completion items

### PR DESCRIPTION
**Pull request by gabro**
Still WIP, but opening for a quick feedback

### CODE REVIEW

File: metaserver/src/test/scala/tests/compiler/CompletionsTest.scala [hunk 1]
```diff
@@ -46,6 +46,7 @@ object CompletionsTest extends CompilerSuite {
       |  "isIncomplete" : false,
       |  "items" : [ {
       |    "label" : "List",
+      |    "kind" : 2,
```
Comment (gabro) [inline metaserver/src/test/scala/tests/compiler/CompletionsTest.scala]: this fails. For some reason `List` symbol is detected as `val`

File: metaserver/src/test/scala/tests/compiler/CompletionsTest.scala [hunk 2]
```diff
@@ -79,6 +80,7 @@ object CompletionsTest extends CompilerSuite {
       |  "isIncomplete" : false,
       |  "items" : [ {
       |    "label" : "StringBuilder",
+      |    "kind" : 2,
```
Comment (gabro) [inline metaserver/src/test/scala/tests/compiler/CompletionsTest.scala]: same as above, this is detected as `val`

File: metaserver/src/test/scala/tests/compiler/CompletionsTest.scala (hunk 1 above)
Comment (olafurpg) [reply metaserver/src/test/scala/tests/compiler/CompletionsTest.scala]: This is correct, from the scala package
```scala
  type List[+A] = scala.collection.immutable.List[A]
  val List = scala.collection.immutable.List
```

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 3]
```diff
@@ -32,11 +33,25 @@ object CompletionProvider extends LazyLogging {
...
+  private def symbolKind(c: Global)(r: c.CompletionResult#M): Option[Int] = {
+    if (r.sym.isConstructor) Some(SymbolKind.Constructor)
+    else if (r.sym.isModuleOrModuleClass) Some(SymbolKind.Module)
+    else if (r.sym.isTraitOrInterface) Some(SymbolKind.Interface)
+    else if (r.sym.isClass) Some(SymbolKind.Class)
+    else if (r.sym.isPackageObject) Some(SymbolKind.Namespace)
+    else if (r.sym.isPackage) Some(SymbolKind.Package)
+    else if (r.sym.isMethod) Some(SymbolKind.Method)
+    else if (r.sym.isVal) Some(SymbolKind.Constant)
+    else if (r.sym.isVar) Some(SymbolKind.Variable)
```
Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Note that `CompletionItemKind` and `SymbolKind` are different things and even same names correspond to different numbers 💣  I think you should use `CompletionItemKind` here. It's already defined [here](https://github.com/scalameta/language-server/blob/master/languageserver/src/main/scala/langserver/types/types.scala#L89), but is by the way incomplete in respect to the v3 of the LSP spec.

Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Argh, thanks, that's a huge mistake I've made. A newtype would've prevented this :(

Comment (laughedelic) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: And I think this is one thing we should change in that code: there's no point to copy the untyped spec style with "named number" if we can have something more Scala-idiomatic (I wanted to say enums, but 😅  you know..)

Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Also, I found https://github.com/Microsoft/language-server-protocol/commit/05e3af34464671a788679080f06558db52d1097b

In version 4 of the protocol we'll be able to specify which symbols we support :)

Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: > And I think this is one thing we should change in that code: there's no point to copy the untyped spec style with "named number" if we can have something more Scala-idiomatic

Yes, and ADT could be a better choice here, or even just a value class.

Comment (laughedelic) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: > In version 4 of the protocol we'll be able to specify which symbols we support :)

I thought it's included in v3 (not released yet), isn't it?

Comment (laughedelic) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Btw, in that referred commit, a special symbol kind for `null` made me laugh :rofl:

**Comment by gabro** [comment]
Ok, I think feature-wise this may be enough to be reviewed/merged.
It's a bit "sad" that `SymbolKind` is richer than `CompletionItemKind` so the icons won't match the one in the code outline exactly... I've found some issues and the LSP seems to be working in that area (also adding new values to those enums), so let's keep an eye on it.

Only thing to notice: I special-cased fields of case classes, so that they stand out (we'll also probably put them at the top in a future PR, see #63):

image

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 4]
```diff
@@ -34,23 +34,23 @@ object CompletionProvider extends LazyLogging {
...
-    else if (r.sym.isVal) Some(SymbolKind.Constant)
-    else if (r.sym.isVar) Some(SymbolKind.Variable)
+  private def completionItemKind(c: Global)(r: c.CompletionResult#M): Option[Int] = {
+    if (r.sym.isPackage) Some(CompletionItemKind.Module)
+    else if (r.sym.isModuleOrModuleClass) Some(CompletionItemKind.Module)
+    else if (r.sym.isTraitOrInterface) Some(CompletionItemKind.Interface)
+    else if (r.sym.isClass) Some(CompletionItemKind.Class)
+    else if (r.sym.isPackageObject) Some(CompletionItemKind.Module)
+    else if (r.sym.isMethod) Some(CompletionItemKind.Method)
+    else if (r.sym.isField && r.sym.owner != null && r.sym.owner.isCaseClass) Some(CompletionItemKind.Field)
```
Comment (laughedelic) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Similarly to this special case, how about using `CompletionItemKind.Property` for vals whose owner is not an object?

Comment (laughedelic) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Alternatively, `Field` could be used for type members (projections) and `Property` for this case.

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 5]
```diff
@@ -32,11 +33,28 @@ object CompletionProvider extends LazyLogging {
...
+  private def completionItemKind(
+      c: Global
+  )(r: c.CompletionResult#M): Option[Int] = {
+    if (r.sym.isPackage) Some(CompletionItemKind.Module)
+    else if (r.sym.isModuleOrModuleClass) Some(CompletionItemKind.Module)
+    else if (r.sym.isTraitOrInterface) Some(CompletionItemKind.Interface)
+    else if (r.sym.isClass) Some(CompletionItemKind.Class)
+    else if (r.sym.isPackageObject) Some(CompletionItemKind.Module)
+    else if (r.sym.isMethod) Some(CompletionItemKind.Method)
+    else if (r.sym.isField && r.sym.owner != null && r.sym.owner.isCaseClass)
```
Comment (olafurpg) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Did you try `.isCaseAccessor` or `sym.isCase && sym.isVal`?

File: metaserver/src/test/scala/tests/compiler/CompletionsTest.scala [hunk 6]
```diff
@@ -108,15 +112,75 @@ object CompletionsTest extends CompilerSuite {
...
+      |  val x: TestTr<<>>
+      |}
+    """.stripMargin,
+    s"""
+       |{
+       |  "isIncomplete" : false,
+       |  "items" : [ {
+       |    "label" : "TestTrait",
+       |    "kind" : ${CompletionItemKind.Interface},
+       |    "detail" : " extends "
```
Comment (olafurpg) [inline metaserver/src/test/scala/tests/compiler/CompletionsTest.scala]: Is this detail expected? Seems a bit weird.

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 7]
```diff
@@ -32,11 +33,28 @@ object CompletionProvider extends LazyLogging {
...
         )
       }
     }
     CompletionList(isIncomplete = false, items = items.result())
   }
 
+  private def completionItemKind(
+      c: Global
+  )(r: c.CompletionResult#M): Option[Int] = {
+    if (r.sym.isPackage) Some(CompletionItemKind.Module)
```
Comment (olafurpg) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Would it make sense to use something else like `Reference` just to distinguish it from objects? Seems like `Namespace` is missing :/

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 8]
```diff
@@ -32,11 +33,28 @@ object CompletionProvider extends LazyLogging {
...
-          detail = Some(r.sym.signatureString)
+          detail = Some(r.sym.signatureString),
+          kind = completionItemKind(compiler)(r)
         )
       }
     }
     CompletionList(isIncomplete = false, items = items.result())
   }
 
+  private def completionItemKind(
```
Comment (olafurpg) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: If you inline this method to `completions` then you avoid the hairy path dependent signature. scala-compiler has lots of nested methods in inside methods.

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala [hunk 9]
```diff
@@ -32,11 +33,28 @@ object CompletionProvider extends LazyLogging {
...
+      c: Global
+  )(r: c.CompletionResult#M): Option[Int] = {
```
Comment (olafurpg) [inline metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: Would be nice if we could to get rid of that `Option[Int]` in favor of a wrapper class 

```scala
case class CompletionItemKind(id: Int)
object CompletionItemKind {
  implicit val writes: Writes[CompletionItemKind] = new Writes[CompletionItemKind] {
    override def writes(o: CompletionItemKind) = JsNumber(o.id)
  }
```

Not blocking in this PR, but maybe in the future.

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala (hunk 7 above)
Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: `Reference` looks very strange on vscode, but I was thinking to simply fallback to `Value` for everything that doesn't have a proper fit in `CompletionItemKind`

File: metaserver/src/test/scala/tests/compiler/CompletionsTest.scala (hunk 6 above)
Comment (gabro) [reply metaserver/src/test/scala/tests/compiler/CompletionsTest.scala]: it's weird but consistent to what I see in the editor. I suspect it's `Template`'s fault. I want to improve details in a future PR (Java LSP seems to be doing a nice job that we can mimic)

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala (hunk 5 above)
Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: ah, no, I haven't tried. Still many `isXXX` methods I'm missing :D

Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: `isCaseAccessor` seems to work!

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala (hunk 7 above)
Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: I think I'll just

Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: I also wouldn't worry too much for now, I'm sure the LSP will evolve in this sense and v4 may include more symbols.

File: metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala (hunk 4 above)
Comment (gabro) [reply metaserver/src/main/scala/scala/meta/languageserver/compiler/CompletionProvider.scala]: I honestly like `Field` for case classes fields. It's quite natural and consistent with my TypeScript experience with VSCode. Can you elaborate on this

> how about using CompletionItemKind.Property for vals whose owner is not an object?

?

**Comment by gabro** [comment]
I addressed @olafurpg's review feedback and defaulted on `Value` for un-handled cases (so that we keep it stable across editors). About the lack of `Namespace` in `CompletionItemKind` I believe the LSP team will improve in that area, so let's just wait and hope.

@laughedelic if you can clarify your comment about using `Property` for vals owned by non-objects I can address it. I'm not sure I understand the case you're talking about. Other than that, I think we can merge this.

**Comment by laughedelic** [comment]
@gabro what I meant is distinguishing "OOP properties" from "FP values" 😄  I don't know if it makes sense in Scala, but if you have something like 

```scala
class Foo(val field: Int) { 
  val property: Int = ??? // or even better var?
}

object Foo {
  val value: Int = ???
}

val foo = new Foo(3)
``` 

We could distinguish these three things: 
* [x] map `Foo.value` to the `Value` kind
* [x] map `foo.field` to the `Field` kind (if I understood it right)
* [ ] map `foo.property` to the `Property` kind 

Probably it makes more sense for `var`s, but my point is that `foo.property` is somewhat like OOP object's property and I think this is what `Property` kind is for (again, probably it's more about mutable object state).

Anyway, this is not critical. So feel free to merge. I just wanted to see if this idea resonates with anybody else.

**Comment by laughedelic** [comment]
> About the lack of Namespace in CompletionItemKind I believe the LSP team will improve in that area, so let's just wait and hope.

Reported here: https://github.com/Microsoft/language-server-protocol/issues/343#issuecomment-347706078

**Comment by olafurpg** [comment]
Nice work @gabro ! Merging to clear queue, excited to see fancy icons in completions :D

**Pull request merged**